### PR TITLE
Add normalizing factors to objectives

### LINKED ISSUE #325: Add normalizing factors to non-dimensionalize objective functions
Right now we compute everything in physical units (Amps, Newtons, Tesla etc). This is good for physics comparison, but makes weighting between different objectives in multi-objective optimization a bit trickier.

In the past we've tried non-dimensionalizing things dynamically by defining a dimensionless metric, but if the normalization is allowed to change during optimization it can cause issues (not to mention making the problem more nonlinear).

Instead we should pre-compute some static scaling factors based on "characteristic scales" in the problem (ie, toroidal flux + area gives a characteristic B field, pressure + radius gives a characteristic force, etc). The goal isn't to be exact, but just to get things all to roughly the same order of magnitude so they can be compared more meaningfully, similar to how we usually plot force error normalized by a pressure gradient.

### PR DESCRIPTION
**Pull request by f0uriest**
- Adds a new function `compute_scaling_factors` to compute characteristic dimensional quantities for an equilibrium
- These are then used to create normalizing scale factors for most objectives
- Should help a bit with poor conditioning in the optimizers, and give dimensionless metrics for comparison.

Resolves #325

### CODE REVIEW

**Comment by f0uriest** [comment]
Would appreciate a sanity check on the units/scaling factors. They seem reasonable to me but there might be better options.

also @unalmis the stability objectives say they're dimensionless, but from the equations it looks like some of them have units? Can you confirm?

**Comment by unalmis** [comment]
Must be a typo in the docstrings/print callback statements.

All the 
- All the Mercier stability critera (D_shear, D_well, D_merc...) are 1/ webers^2.
    - `D_shear` is just (diota/dpsi)^2  = 1 / webers^2.
    - `D_Mercier` is the sum of the other 4 stability quantities
- `magnetic well ` is fine (dimensionless)

**Comment by dpanici** [comment]
What sorts of scale do the normalized quantities get? Are the normalized values relatively consistent across equilibria?

What you've chosen for each seems reasonable to me though

**Review by dpanici — DISMISSED** [review]

File: desc/objectives/_qs.py [hunk 1]
```diff
@@ -253,25 +280,37 @@ class QuasisymmetryTwoTerm(_Objective):
...
+            target=target,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            name=name,
+        )
+
         self._print_value_fmt = (
-            "Quasi-symmetry ({},{}) error: ".format(self.helicity[0], self.helicity[1])
+            "Quasi-symmetry two-term ({},{}) error: ".format(
```
Comment (ddudt) [inline desc/objectives/_qs.py:310]: Should be `"Quasi-symmetry ({},{}) two-term error: "` if we want to follow the same formatting as the Boozer objective.

**Review by ddudt — CHANGES_REQUESTED** [review]
1. We should also add the normalization options to the linear objectives, and make it part of the objectives API in the `_Objective` class. In typical usage the linear objectives are always satisfied exactly so the normalizations won't matter, but we should allow more general usage. 
2. Do we have any evidence that these normalizations actually improve the optimizations? The default is set to `normalize=True`, but none of the tests use it. **I will veto changing the default behavior until we are confident that it is usually an improvement.** And either way, we should include optimization tests that use the normalizations.

File: desc/objectives/normalization.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    """Compute dimensional quantities for normalizations."""
+    scales = {}
+    R10 = eq.Rb_lmn[eq.surface.R_basis.get_idx(M=1, N=0)]
+    Z10 = eq.Zb_lmn[eq.surface.Z_basis.get_idx(M=-1, N=0)]
+    R00 = eq.Rb_lmn[eq.surface.R_basis.get_idx(M=0, N=0)]
+
+    scales["R0"] = R00
+    scales["a"] = np.sqrt(np.abs(R10 * Z10))
+    scales["V"] = 2 * np.pi * scales["R0"] * scales["a"]
+    scales["A"] = np.pi * scales["a"] ** 2
```
[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/normalization.py:17]: Yeah I just wanted to keep it simple and not have to deal with transforming to real space. I checked several equilibria and these approximations agree within a few percent.

File: desc/objectives/normalization.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    scales["B_T"] = eq.Psi / scales["A"]
```
[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/normalization.py:18]: yeah ill fix that

File: desc/objectives/normalization.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    iota = eq.get_profile("iota")(np.linspace(0, 1, 20))
+    scales["B_P"] = scales["B_T"] * np.mean(np.abs(iota))
+    scales["B"] = np.sqrt(scales["B_T"] ** 2 + scales["B_P"] ** 2)
+    scales["I"] = scales["B_P"] / (2 * np.pi * mu_0)
```
[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/normalization.py:22]: yep

File: desc/objectives/_qs.py (hunk 1 above)
Comment (f0uriest) [reply desc/objectives/_qs.py:310]: fixed

**Comment by f0uriest** [comment]
I think it's hard to say that they help the optimization, since effectively they're redundant with ``weight``, but the idea is that they make it easier to choose ``weight`` without having to think about the scales of different objectives. So it's entirely a UI improvement, it doesn't change any of the actual physics. But I see your point. I ran some examples locally that seem fine, I just didn't use normalization in any of the tests because then we'd have to mess with the weights to get the same result.

**Comment by f0uriest** [comment]
Ultimately to get "good" results we'll still have to play around with the weights, the hope is just that this makes that a bit easier and more intuitive.

**Comment by ddudt** [comment]
> I think it's hard to say that they help the optimization, since effectively they're redundant with `weight`, but the idea is that they make it easier to choose `weight` without having to think about the scales of different objectives. So it's entirely a UI improvement, it doesn't change any of the actual physics. But I see your point. I ran some examples locally that seem fine, I just didn't use normalization in any of the tests because then we'd have to mess with the weights to get the same result.

OK, I see your point too. I still think we should have a test somewhere that uses normalization so those new lines of code are hit. 

And what do you think about my first review comment of adding the normalization arguments to the `_Objectives` API?

**Comment by f0uriest** [comment]
Yeah I think that makes sense, I just wasn't sure since some of the objectives are already dimensionless, but I guess we can just default to ``scale=1`` in that case. I'll try to add it to the rest of the objectives today.

File: desc/objectives/utils.py [hunk 5]
```diff
@@ -153,14 +166,16 @@ def factorize_linear_constraints(constraints, objective_args):
         else:
             unfixed_args.append(arg)
             A_ = obj.derivatives["jac"][arg](jnp.zeros(obj.dimensions[arg]))
-            b_ = obj.target
+            # using obj.compute instead of obj.target to allow for correct scale/weight
+            b_ = -obj.compute(jnp.zeros(obj.dimensions[arg]))
```
Comment (dpanici) [inline desc/objectives/utils.py:170]: why is there a negative sign here?

File: desc/objectives/utils.py [hunk 6]
```diff
@@ -153,14 +166,16 @@ def factorize_linear_constraints(constraints, objective_args):
...
             if A_.shape[0]:
                 Ainv_, Z_ = svd_inv_null(A_)
             else:
                 Ainv_ = A_.T
             A[arg] = A_
             b[arg] = b_
-            Ainv[arg] = Ainv_
+            # need to undo scaling here to work with perturbations
+            Ainv[arg] = Ainv_ * obj.weight / obj.normalization
```
Comment (dpanici) [inline desc/objectives/utils.py:179]: why is obj.weight multiplied here? this is related to the bug you had mentioned about it not working correctly when the objective is weighted?

File: desc/objectives/_wrappers.py [hunk 7]
```diff
@@ -229,27 +230,19 @@ def jac(self, x):
...
-        }
-        Gx = {arg: Gx[:, self._objective.x_idx[arg]] for arg in self._objective.args}
-        for arg in self._eq_objective.args:
-            if arg not in Fx.keys():
-                Fx[arg] = jnp.zeros((self._eq_objective.dim_f, self.dimensions[arg]))
-            if arg not in Gx.keys():
-                Gx[arg] = jnp.zeros((self._objective.dim_f, self.dimensions[arg]))
-        Fx = jnp.hstack([Fx[arg] for arg in arg_order if arg in Fx])
-        Gx = jnp.hstack([Gx[arg] for arg in arg_order if arg in Gx])
+        Fx = align_jacobian(Fx, self._eq_objective, self._objective)
```
Comment (dpanici) [inline desc/objectives/_wrappers.py:233]: this is just a refactor of the old code?

File: desc/objectives/utils.py (hunk 5 above)
Comment (f0uriest) [reply desc/objectives/utils.py:170]: The residual is computed as f(x) = Ax-b, so f(0) = -b, so we need the negative sign to get the correct b

File: desc/objectives/utils.py (hunk 6 above)
Comment (f0uriest) [reply desc/objectives/utils.py:179]: Yeah, `Ainv` here is only used as the relation between R and Rb, eg dR/dRb = Ainv, so we want that to be the unweighted version.

A R = Rb
(w A) R = w Rb
dR/dRb = inv(w A) * w

File: desc/objectives/_wrappers.py (hunk 7 above)
Comment (f0uriest) [reply desc/objectives/_wrappers.py:233]: yeah we were doing this in a few places and will probably need to also do it for more multiobjective/constrained stuff so made sense to just roll it into a utility function.

File: desc/objectives/_stability.py [hunk 8]
```diff
@@ -184,7 +182,9 @@ def compute(self, R_lmn, Z_lmn, L_lmn, p_l, i_l, c_l, Psi, **kwargs):
             self._iota,
             self._current,
         )
-        return self._shift_scale(compress(self.grid, data["D_Mercier"]))
+        f = compress(self.grid, data["D_Mercier"], surface_label="rho")
+        w = compress(self.grid, self.grid.weights, surface_label="rho")
+        return self._shift_scale(f * w)
```
Comment (ddudt) [inline desc/objectives/_stability.py:187]: In the `ToroidalCurrent` objective the radial weights are computed as follows:
```
I = compress(self.grid, data["current"], surface_label="rho")
w = compress(self.grid, self.grid.spacing[:, 0], surface_label="rho")
return self._shift_scale(I * w)
```
Why would it be any different for the Mercier objectives? I mean the difference between using `self.grid.weights` vs `self.grid.spacing[:, 0]`

File: desc/objectives/_stability.py [hunk 9]
```diff
@@ -329,4 +353,6 @@ def compute(self, R_lmn, Z_lmn, L_lmn, p_l, i_l, c_l, Psi, **kwargs):
             self._iota,
             self._current,
         )
-        return self._shift_scale(compress(self.grid, data["magnetic well"]))
+        f = compress(self.grid, data["magnetic well"], surface_label="rho")
+        w = compress(self.grid, self.grid.weights, surface_label="rho")
```
Comment (ddudt) [inline desc/objectives/_stability.py:357]: see previous comment

File: tests/test_objective_funs.py [hunk 10]
```diff
@@ -93,6 +96,7 @@ def test(eq):
             W = obj.compute(
                 eq.R_lmn, eq.Z_lmn, eq.L_lmn, eq.p_l, eq.i_l, eq.c_l, eq.Psi
             )
```
Comment (ddudt) [inline tests/test_objective_funs.py:98]: Looks like we can delete these lines since they are now redundant with line 99.

**Review by ddudt — CHANGES_REQUESTED** [review]

File: tests/test_objective_funs.py (hunk 10 above)
Comment (f0uriest) [reply tests/test_objective_funs.py]: yep, done

**Review by dpanici — APPROVED** [review]
